### PR TITLE
[Parse] [Sema] [SR-3174] Ignore local variables in selector

### PR DESCRIPTION
**Pull request by rmnblm**
Declaring a variable and assigning a selector that refers to a method with the same name as the variable is currently not possible. The compiler will complain with an error message.

``` swift
func foo() {
  let bar = #selector(bar) // Error: Variable used within its own initial value
}

func bar() { }
```

However, the compiler should realise that not the variable, but instead, the method `bar()` was meant to be used. A local variable can't even be a selector.

> The subexpression of the #selector expression must be a reference to an objc method. [SE-0022](https://github.com/apple/swift-evolution/blob/master/proposals/0022-objc-selectors.md#detailed-design)

Again, the compiler will complain with an appropriate error message.

``` swift 
func foo() {
  let t = { 
    print("Swift is awesome!")
  }
  let bar = #selector(t) // Error: Argument of '#selector' cannot refer to variable 't'
}
```

This pull request changes this behaviour and allows the usage of the same name in a selector.

``` swift
func foo() {
  let bar = #selector(bar) // All good, no error anymore
}

func bar() { }
```

The compiler does this now in two steps: Firstly, the parser ignores the `DisabledVars` property if a selector is currently being parsed. In addition, while the type checker walks the AST and stumbles across an `ObjCSelectorExpr`, it ignores local variables when trying to resolve an `UnresolvedDeclRefExpr`.

Resolves [SR-3174](https://bugs.swift.org/browse/SR-3174).

Can you take a look at it, @DougGregor or @jrose-apple? 🙏

### CODE REVIEW

File: lib/Sema/TypeCheckConstraints.cpp [hunk 1]
```diff
@@ -813,6 +815,15 @@ namespace {
         return finish(true, TC.resolveDeclRefExpr(unresolved, DC));
       }
 
+      if (auto selector = dyn_cast<ObjCSelectorExpr>(expr)) {
```
Comment (CodaFi) [inline lib/Sema/TypeCheckConstraints.cpp]: Can you leave a short comment explaining why selector references are exempt from circularity checks.

File: test/expr/unary/selector/selector.swift [hunk 2]
```diff
@@ -146,3 +146,19 @@ case #selector(SR1827.bar)?:
 default:
   break
 }
+
+// SR-3174
+
+class Hello {
```
Comment (CodaFi) [inline test/expr/unary/selector/selector.swift]: In addition to these tests, please add one with nested declarations and another that uses a top-level selector.

File: lib/Sema/TypeCheckConstraints.cpp (hunk 1 above)
Comment (rmnblm) [inline lib/Sema/TypeCheckConstraints.cpp]: I just added a short comment in the code (see latest commit).

File: test/expr/unary/selector/selector.swift (hunk 2 above)
Comment (rmnblm) [inline test/expr/unary/selector/selector.swift]: I added a test for a nested declaration. What exactly do you mean by "top-level selector"?

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by rmnblm** [comment]
@CodaFi I added another test, you can smoke test again.

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckConstraints.cpp [hunk 3]
```diff
@@ -813,6 +815,18 @@ namespace {
         return finish(true, TC.resolveDeclRefExpr(unresolved, DC));
       }
 
+      // For method selectors with unresolved sub-expressions,
+      // resolve the expression but ignore local variables to allow
+      // conflicting method names
+      if (auto selector = dyn_cast<ObjCSelectorExpr>(expr)) {
+        if (selector->isMethodSelector()) {
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckConstraints.cpp]: What does being a method have to do with ignoring local variables?

**Review by jrose-apple — COMMENTED** [review]
Thanks for working on this! This is what I called the "basic fix", which disallows this currently-legal code:

```swift
import Foundation
func test(_ obj: NSObject) {
  let x = #selector(obj.copy)
  print(x)
}
test(NSObject())
```

As such, we'd need to limit the change to Swift 4 mode. You can test `TC.getLangOptions().isSwiftVersion3()` for that.

File: lib/Sema/TypeCheckConstraints.cpp (hunk 3 above)
Comment (rmnblm) [inline lib/Sema/TypeCheckConstraints.cpp]: Well, correct me if I'm wrong but if it's a getter/setter selector, ignoring local variables doesn't make sense so there is only the method selector left where conflicts can occur with local variable names.

Comment (rmnblm) [inline lib/Sema/TypeCheckConstraints.cpp]: To limit the change to Swift 4, should I wrap lines 821 to 828 with, would this be sufficient?

``` c++
if !TC.getLangOptions().isSwiftVersion3() { 
  ...
}
```

Comment (jrose-apple) [inline lib/Sema/TypeCheckConstraints.cpp]: Local variables still won't have selectors.

```swift
import Foundation
class Foo: NSObject {
  @objc var value: Int = 0
  func test() {
    var value = 42
    print(#selector(getter:value))
  }
}
```

Comment (jrose-apple) [inline lib/Sema/TypeCheckConstraints.cpp]: And yes, that seems like the right way to limit it to Swift 4.

**Comment by jrose-apple** [comment]
@DougGregor ping

**Comment by rmnblm** [comment]
@jrose-apple By the way, I'm still on it and will fix your mentioned issue that this doesn't work anymore: 

``` swift 
import Foundation
func test(_ obj: NSObject) {
  let x = #selector(obj.copy)
  print(x)
}
test(NSObject())
```

Unfortunately, there's a lot going on right now and time's rare. I hope I can finish this PR.

**Comment by rmnblm** [comment]
@jrose-apple I tried your currently-legal code that is now disallowed due to this PR (copied it to the end of the file `test/expr/unary/selector/selector.swift`) on the master branch and it gives me the same error, am I missing something here? (The code runs fine in a playground.)

In the screenshot below you can see the command I executed.

screen shot 2017-05-25 at 23 01 55

**Comment by jrose-apple** [comment]
Ah, that test probably runs against our fake version of the Mac SDK (in test/Inputs/clang-importer-sdk/), which wouldn't have all of NSObject's methods declared. It doesn't really matter which method you pick here; the point is it's accessing something on a local variable.

**Review by shahzadlone — APPROVED** [review]

**Comment by CodaFi** [comment]
@rmnblm I'm going to close this out since it's been two years, but I think the idea you've got is correct and can probably be retailored to fix this SR.  If you're interested in picking this back up, either reopen this pull request after a rebase or open a new one.  I'm also resetting the assignee for this SR so somebody else can take a look.

Thanks for the work you've done so far.  Hopefully it can help another contributor out in the future.

**Pull request closed without merging**